pinot-core/src/test/java/org/apache/pinot/queries/CovarianceQueriesTest.java

```diff
@@ -0,0 +1,444 @@
+      _sumDoubleXY += doubleX * doubleY;
+      _sumIntDouble += intX * doubleX;
+      _sumIntLong += intX * longVal;
+      _sumIntFloat += intX * floatCol[i];
+      _sumDoubleLong += doubleX * longVal;
+      _sumDoubleFloat += doubleX * floatCol[i];
+      _sumLongFloat += longVal * floatCol[i];
+
+      record.putValue(INT_COLUMN_X, intX);
+      record.putValue(INT_COLUMN_Y, intY);
+      record.putValue(DOUBLE_COLUMN_X, doubleX);
+      record.putValue(DOUBLE_COLUMN_Y, doubleY);
+      record.putValue(LONG_COLUMN, longVal);
+      record.putValue(FLOAT_COLUMN, floatVal);
+      record.putValue(GROUP_BY_COLUMN, groupByVal);
+      records.add(record);
+    }
+    _expectedGroupByResultVer1[groupByVal] = new CovarianceTuple(sumX, sumGroupBy, sumXGroupBy, groupSize);
+    _expectedGroupByResultVer2[groupByVal] = new CovarianceTuple(sumX, sumY, sumXY, groupSize);
+
+    // calculate inter segment result
+    Covariance cov = new Covariance();
+    double[] newIntColX = Arrays.stream(intColX).asDoubleStream().toArray();
+    double[] newIntColY = Arrays.stream(intColY).asDoubleStream().toArray();
+    double[] newLongCol = Arrays.stream(longCol).asDoubleStream().toArray();
+    _expectedCovIntXY = cov.covariance(newIntColX, newIntColY, false);
+    _expectedCovDoubleXY = cov.covariance(doubleColX, doubleColY, false);
+    _expectedCovIntDouble = cov.covariance(newIntColX, doubleColX, false);
+    _expectedCovIntLong = cov.covariance(newIntColX, newLongCol, false);
+    _expectedCovIntFloat = cov.covariance(newIntColX, floatCol, false);
+    _expectedCovDoubleLong = cov.covariance(doubleColX, newLongCol, false);
+    _expectedCovDoubleFloat = cov.covariance(doubleColX, floatCol, false);
+    _expectedCovLongFloat = cov.covariance(newLongCol, floatCol, false);
+
+    double[] filteredX = Arrays.copyOfRange(doubleColX, 0, NUM_RECORDS / 2);
+    double[] filteredY = Arrays.copyOfRange(doubleColY, 0, NUM_RECORDS / 2);
+    _expectedCovWithFilter = cov.covariance(filteredX, filteredY, false);
+
+    // calculate inter segment group by results
+    for (int i = 0; i < NUM_GROUPS; i++) {
+      double[] colX = Arrays.copyOfRange(doubleColX, i * groupSize, (i + 1) * groupSize);
+      double[] colGroupBy = Arrays.copyOfRange(groupByCol, i * groupSize, (i + 1) * groupSize);
+      double[] colY = Arrays.copyOfRange(doubleColY, i * groupSize, (i + 1) * groupSize);
+      _expectedFinalResultVer1[i] = cov.covariance(colX, colGroupBy, false);
+      _expectedFinalResultVer2[i] = cov.covariance(colX, colY, false);
+    }
+
+    // generate testSegment
+    ImmutableSegment immutableSegment = setUpSingleSegment(records, SEGMENT_NAME);
+    _indexSegment = immutableSegment;
+    _indexSegments = Arrays.asList(immutableSegment, immutableSegment);
+
+    // divide testSegment into 4 distinct segments for distinct inter segment tests
+    // by doing so, we can avoid calculating global covariance again
+    _instances = new ArrayList<>();
```

Comment:
We can probably create an 

1. add function `getDistinctInstances` in baseQueriesTest and pass this in
2. or add a new code path in `getBrokerResponse` if we pass 4 segments in `getIndexSegments` (combining `getBrokerResponseDistinctInstance` with `getBrokerResponse`)

discussed offline